resources/test-runner.mjs

```diff
@@ -0,0 +1,96 @@
+import { TEST_INVOKER_LOOKUP } from "./test-invoker.mjs";
+
+export class TestRunner {
+    constructor(suite, test, params, callback, page, frame) {
+        this._suite = suite;
+        this._test = test;
+        this._params = params;
+        this._callback = callback;
+
+        this._page = page;
+        this._frame = frame;
+    }
+
+    async runTest() {
+        // Prepare all mark labels outside the measuring loop.
+        const suiteName = this._suite.name;
+        const testName = this._test.name;
+        const syncStartLabel = `${suiteName}.${testName}-start`;
+        const syncEndLabel = `${suiteName}.${testName}-sync-end`;
+        const asyncStartLabel = `${suiteName}.${testName}-async-start`;
+        const asyncEndLabel = `${suiteName}.${testName}-async-end`;
+
+        let syncTime;
+        let asyncStartTime;
+        let asyncTime;
+        const runSync = () => {
+            this._runWarmupSuite();
+            syncTime = this._measureSyncTime(syncStartLabel, syncEndLabel);
+            asyncStartTime = this._initAsyncTime(asyncStartLabel);
+        };
+        const measureAsync = () => {
+            this._forceLayout();
+            asyncTime = this._measureAsyncTime(asyncStartTime, asyncEndLabel);
+            this._measureTestPerformance(suiteName, testName, syncStartLabel, syncEndLabel, asyncStartLabel, asyncEndLabel);
+        };
+
+        const report = () => this._callback(this._test, syncTime, asyncTime);
+        const invokerClass = TEST_INVOKER_LOOKUP[this._params.measurementMethod];
+        const invoker = new invokerClass(runSync, measureAsync, report, this._params);
+
+        return invoker.start();
+    }
+
+    _runWarmupSuite() {
```

Comment:
This isn't really a suite. We should probably just call it _runWarmup or something

Reply:
done